### DISCUSSION TITLE
Oqtane, Telerik Reporting and System.Data.SqlClient vs. Microsoft.Data.SqlClient

### DISCUSSION THREAD

**Discussion by nicpitsch** [oqtane/oqtane.framework#1929] [Q&A] [answered]
*Telerik Reporting* uses *System.Data.SqlClient*. A sample Configuration based on [Telerik Docs](https://docs.telerik.com/reporting/telerik-reporting-rest-service-aspnetcore-net5#add-configuration-settings-optional)..

```json
{
  "ConnectionStrings": {
    "Telerik.Reporting.ConnectionString": {
      "connectionString": "Data Source=.\\SQLEXPRESS;Initial Catalog=MyOqtaneDB;Integrated Security=true",
      "providerName": "System.Data.SqlClient"
    }
  }
}
```

..where *MyOqtaneDB* would be the Oqtane based Sql Server Database.

Problem:

I add a configuration layer with *Telerik.Reporting.ConnectionString* in Startup.cs of my custom module. Oqtane reads it at first and raises an error `Unable to establish a connection to the database. Please verify that your connection string is valid. In case you use a named connection string from the application configuration file, make sure the name is correct and the connection string settings are present in the configuration file of your application`. 

If I change the ConnectionString name to *DefaultConnection* the error is:

```ini
The type initializer for 'System.Data.SqlClient.TdsParser' threw an exception.
------------- InnerException -------------
The type initializer for 'System.Data.SqlClient.SNILoadHandle' threw an exception.
------------- InnerException -------------
Unable to load DLL 'sni.dll' or one of its dependencies
```

In my understanding that's because Oqtane uses *Microsoft.Data.SqlClient*.

Any hint how one could 'hide' the *Telerik.Reporting.ConnectionString* against Oqtane or an alternative solution? 
This issue seems similar to #1770. Maybe @mixlplix?

**Comment by hishamco** [comment]
Seems this by design

https://github.com/oqtane/oqtane.framework/blob/d093c03d929dfd85c9f450a6b22dbce24502fb21/Oqtane.Server/Repository/Context/MasterDBContext.cs#L45

https://github.com/oqtane/oqtane.framework/blob/6eaa3d259ecef6bd167313c93b1c0260b3117ace/Oqtane.Server/Infrastructure/Middleware/TenantMiddleware.cs#L20

/cc @sbwalker

**Comment by sbwalker** [comment]
@nicpitsch Oqtane only relies on a single connection string by convention named "DefaultConnection" which must be declared in appsettings.json - it does not look at any other database connections. Oqtane supports multiple types of databases so this connection string does NOT relate to Microsoft.Data.SqlClient - it relates to whatever database is specified as the DefaultDBType in appsettings.json - which could be SQL Server, MySQL, SQLite, PostgreSQL. The connection string does not specify the providerName attribute (it is not necessary). If you want to use Telerik reporting then it appears you may need to specify an additional connection string in appsettings.json (I am guessing your logic which instantiates the reporting service is somehow referencing the "Telerik.Reporting.ConnectionString" by name in code?):

```
  "ConnectionStrings": {
    "DefaultConnection": "{Your Oqtane Connection String}",
    "Telerik.Reporting.ConnectionString": "Data Source=.\\SQLEXPRESS;Initial Catalog=MyOqtaneDB;Integrated Security=true"
  },

```
Keep in mind that Oqtane is multi-tenant. The DefaultConnection only refers to the "master" database in the installation. If you are using a different database for each of your sites (and not a single shared database), then these additional connections are stored in the Tenant table - DBConnectionString column within the master database. Obviously this will create some additional techical challenges if you are leveraging this capability.

**Reply by nicpitsch** [reply]
> If you are using a different database for each of your sites (and not a single shared database)

No, I'm only using one database. I thought the issue would be related somehow with the connection string format, therefore i tried with a second, specific connection string for Telerik Reporting.

**Comment by nicpitsch** [comment]
Thanks @hishamco and @sbwalker for the responses and hints.

I went a bit in the wrong direction. Forget the  configuration layer with *Telerik.Reporting.ConnectionString*. The *DefaultConnection* is fine for Telerik Reporting.

What i tried next: i renamed *System.Data.SqlClient.dll* to *System.Data.SqlClient.Oqtane.dll* based on response in [issue #1624](https://github.com/oqtane/oqtane.framework/discussions/1624#discussioncomment-1244300). As the Debug output shows, the Oqtane method *LoadAssemblies* indeed loads the Assembly System.Data.SqlClient. But this error still raises:

```ini
The type initializer for 'System.Data.SqlClient.TdsParser' threw an exception.
------------- InnerException -------------
The type initializer for 'System.Data.SqlClient.SNILoadHandle' threw an exception.
------------- InnerException -------------
Unable to load DLL 'sni.dll' or one of its dependencies
```

Provisional solution: i installed *System.Data.SqlClient* package in *Oqtane.Server*. Now *Telerik Reporting* is functional! But this isn't an option with regard to deployment to a production environment. @sbwalker would it make sense to include *System.Data.SqlClient* package in Oqtane? Or do you see another solution? 

As it seems *BoldReports* is struggling with the same problem (issue #1770).

**Reply by sbwalker** [reply]
For performance and security reasons, Oqtane does not include any dependencies unless there are required by the core framework. The core framework does not use System.Data.SqlClient therefore the package should not be included.

However, it sounds like Telerik Reporting does have a dependency on System.Data.SqlClient. And based on the error message I am also guessing that System.Data.SqlClient most likely has dependencies on other DLLs. The reason why this works when you add a package reference to Oqtane Server is because it installs ALL of the DLL dependencies into the /bin. When you are developing a custom Oqtane module the DLLs are not automatically deployed to the /bin - they need to be explicitly identified in the Package/debug.cmd for your module. 

See how RestSharp is deployed below in the DNF.Projects example ( luckily RestSharp has no other dependencies but if it did then all of them would need to be explicitly declared as well )

```
XCOPY "..\Client\bin\Debug\net6.0\DNF.Projects.Client.Oqtane.dll" "..\..\oqtane.framework\Oqtane.Server\bin\Debug\net6.0\" /Y
XCOPY "..\Client\bin\Debug\net6.0\DNF.Projects.Client.Oqtane.pdb" "..\..\oqtane.framework\Oqtane.Server\bin\Debug\net6.0\" /Y
XCOPY "..\Server\bin\Debug\net6.0\DNF.Projects.Server.Oqtane.dll" "..\..\oqtane.framework\Oqtane.Server\bin\Debug\net6.0\" /Y
XCOPY "..\Server\bin\Debug\net6.0\DNF.Projects.Server.Oqtane.pdb" "..\..\oqtane.framework\Oqtane.Server\bin\Debug\net6.0\" /Y
XCOPY "..\Server\bin\Debug\net6.0\RestSharp.dll" "..\..\oqtane.framework\Oqtane.Server\bin\Debug\net6.0\" /Y
XCOPY "..\Shared\bin\Debug\net6.0\DNF.Projects.Shared.Oqtane.dll" "..\..\oqtane.framework\Oqtane.Server\bin\Debug\net6.0\" /Y
XCOPY "..\Shared\bin\Debug\net6.0\DNF.Projects.Shared.Oqtane.pdb" "..\..\oqtane.framework\Oqtane.Server\bin\Debug\net6.0\" /Y
XCOPY "..\Server\wwwroot\Modules\DNF.Projects\*" "..\..\oqtane.framework\Oqtane.Server\wwwroot\Modules\DNF.Projects\" /Y /S /I

```

**Comment by todorarabadzhiev** [comment]
Hi @nicpitsch and @sbwalker.
In Telerik Reporting, you may utilize any ADO.NET data provider to connect your reports to a database. Our code relies entirely on the data provider for connecting and fetching the schema and the data from the database. If the provider is not specified, we assume that you would like to use the System.Data.SqlClient. I guess this explains the observed error.

@nicpitsch, you need to specify the data provider in the connection string. The Microsoft.Data.SqlClient provider is not supported out of the box, hence, you need to register it as explained in the KB article [How to register a DbProviderFactory in a .NET Core project](https://docs.telerik.com/reporting/knowledge-base/how-to-register-db-provider-factory-in-net-core-project).

**Comment by nicpitsch** [comment]
Thanks @todorarabadzhiev for the support. The intention is to get rid of *System.Data.SqlClient* data provider and to replace it with *Microsoft.Data.SqlClient* concerning Telerik Reporting.

*ReportsController.cs* setup (in Oqtane custom module *my.Module.Server*):

```c#
namespace my.Module.Controllers
{
    /// <summary>
    /// Setup Telerik Reporting REST service to expose the Reports Generation Engine to an ASP.NET Core in .NET >=5 Web Application 
    /// (https://docs.telerik.com/reporting/telerik-reporting-rest-service-aspnetcore-net5#setting-up-the-rest-service).
    /// </summary>
    [Route("api/reports")]
    [ApiController]
    public class ReportsController : ReportsControllerBase
    {
        public ReportsController(IReportServiceConfiguration reportServiceConfiguration)
            : base(reportServiceConfiguration)
        {
            Telerik.Reporting.Processing.Data.DbProviderFactories.RegisterFactory("Microsoft.Data.SqlClient", Microsoft.Data.SqlClient.SqlClientFactory.Instance);
        }
    }
}
```

Connection string setup using a further configuration layer (in *appsettings.reporting.json*). Remark: my Oqtane solution has (only) one master database. This setup is applied to accomplish the Telerik conditions (see below):

```json
{
  "ConnectionStrings": {
    "Telerik.Reporting.ConnectionString": {
      "connectionString": "Data Source=.\\SQLEXPRESS;Initial Catalog=MyOqtaneDB;Integrated Security=true",
      "providerName": "Microsoft.Data.SqlClient"
    }
  }
}
```

It's still not functional, the error output in Oqtane.Server Debug window is `DataSource connection string: Telerik.Reporting.ConnectionString; Resolved connection string: Telerik.Reporting.ConnectionString; Exception: System.ArgumentException: Format of the initialization string does not conform to specification starting at index 0.`. 

@sbwalker is this an Oqtane issue? Notice that i have to setup the connection string format using the *providerName* according to the 2nd example in [Telerik Add Configuration Settings (Optional)](https://docs.telerik.com/reporting/telerik-reporting-rest-service-aspnetcore-net5#add-configuration-settings-optional) based on Todor's response.  Notice the different format compared to Oqtane's *DefaultConnection* setup in appsettings.json (https://github.com/oqtane/oqtane.framework/commit/b21ffc2d6323d1c2e45fc11c63b09f5b469afcf2).

**Reply by sbwalker** [reply]
@nicpitsch No this is not an Oqtane issue - it is a Telerik integration issue. As I mentioned above Oqtane relies on a connection string named DefaultConnection. My suggestion is that you should define a separate connection for Telerik - even if it points to the same database:

```
  "ConnectionStrings": {
    "DefaultConnection": "{Your Oqtane Connection String}",
    "Telerik.Reporting.ConnectionString": {
        "connectionString": "{Your Oqtane Connection String}",
        "providerName": "Microsoft.Data.SqlClient"
    }
  },
```

**Comment by nicpitsch** [comment, marked as answer]
> @nicpitsch No this is not an Oqtane issue - it is a Telerik integration issue. As I mentioned above Oqtane relies on a connection string named DefaultConnection. My suggestion is that you should define a separate connection for Telerik - even if it points to the same database:
> 
> ```
>   "ConnectionStrings": {
>     "DefaultConnection": "{Your Oqtane Connection String}",
>     "Telerik.Reporting.ConnectionString": {
>         "connectionString": "{Your Oqtane Connection String}",
>         "providerName": "Microsoft.Data.SqlClient"
>     }
>   },
> ```

@sbwalker that's it! Your ConnectionString format (implemented in appsettings.json only) ist the solution. Thank you and @todorarabadzhiev very much!

As an overview information for other *Oqtane* and *Telerik* users/developers. This issue is regarding the integration of *Telerik Reporting* using a [Blazor Report Viewer](https://docs.telerik.com/reporting/blazor-report-viewer-how-to-use). It concerns following setup:

* Oqtane custom ([monolithic](https://github.com/oqtane/oqtane.framework/discussions/689)) Module based on Oqtane Blazor Server hosting model,
* with Telerik Reporting for .NET Core using the [*SqlDataSource*](https://docs.telerik.com/reporting/sqldatasource) Component for SQL Sever.
* It implements the *Microsoft.Data.SqlClient* data provider based on the *ReportsController* controller class setup [below](https://github.com/oqtane/oqtane.framework/discussions/1929#discussioncomment-1967442). In my experience it is essential to **not** use the *System.Data.SqlClient* data provider if you don't want to fail sooner or later.

Future: considering the Oqtane environment it would probably be more obvious to implement the [*WebServiceDataSource*](https://docs.telerik.com/reporting/webservicedatasource-component) Component instead. As for now i didn't find a way how to authenticate  the Telerik *WebServiceDataSource* access against Oqtane.

> Hi @nicpitsch, Just one more recommendation. You register _Microsoft.Data.SqlClient_ provider in the constructor of the ReportsController. Note that this may be an overhead as a ReportsController instance gets created on each request to the Telerik Reporting REST Service. For that reason, we recommend doing this in a static constructor that would be called only once in the application lifetime. Here is the updated code I suggest :
> 
> ```cs
> namespace my.Module.Controllers
> {
>     /// <summary>
>     /// Setup Telerik Reporting REST service to expose the Reports Generation Engine to an ASP.NET Core in .NET >=5 Web Application 
>     /// (https://docs.telerik.com/reporting/telerik-reporting-rest-service-aspnetcore-net5#setting-up-the-rest-service).
>     /// </summary>
>     [Route("api/reports")]
>     [ApiController]
>     public class ReportsController : ReportsControllerBase
>     {
> 	static ReportsController()
> 	{
> 		Telerik.Reporting.Processing.Data.DbProviderFactories.RegisterFactory("Microsoft.Data.SqlClient", Microsoft.Data.SqlClient.SqlClientFactory.Instance);
> 	}
> 	
>         public ReportsController(IReportServiceConfiguration reportServiceConfiguration)
>             : base(reportServiceConfiguration)
>         {
>             
>         }
>     }
> }
> ```

**Comment by todorarabadzhiev** [comment]
Hi @nicpitsch,
Just one more recommendation.
You register _Microsoft.Data.SqlClient_ provider in the constructor of the ReportsController. Note that this may be an overhead as a ReportsController instance gets created on each request to the Telerik Reporting REST Service. For that reason, we recommend doing this in a static constructor that would be called only once in the application lifetime. Here is the updated code I suggest :

```C#
namespace my.Module.Controllers
{
    /// <summary>
    /// Setup Telerik Reporting REST service to expose the Reports Generation Engine to an ASP.NET Core in .NET >=5 Web Application 
    /// (https://docs.telerik.com/reporting/telerik-reporting-rest-service-aspnetcore-net5#setting-up-the-rest-service).
    /// </summary>
    [Route("api/reports")]
    [ApiController]
    public class ReportsController : ReportsControllerBase
    {
	static ReportsController()
	{
		Telerik.Reporting.Processing.Data.DbProviderFactories.RegisterFactory("Microsoft.Data.SqlClient", Microsoft.Data.SqlClient.SqlClientFactory.Instance);
	}
	
        public ReportsController(IReportServiceConfiguration reportServiceConfiguration)
            : base(reportServiceConfiguration)
        {
            
        }
    }
}
```
